after upgrade to 20 tests fail

### Command

update

### Is this a regression?

- [x] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

19

### Description

1. After running `ng update @angular/core@20 @angular/cli@20` 
2. Then `npm run test` I get error

```
 Error: Cannot find module '@angular-devkit/build-angular/plugins/karma'
```

### Minimal Reproduction

npm run test

### Exception or Error

```text

```

### Your Environment

```text
Angular CLI: 20.0.0
Node: 22.15.1
Package Manager: npm 10.9.2
OS: win32 x64

Angular: 20.0.0
... animations, build, cli, common, compiler, compiler-cli, core
... forms, language-service, localize, platform-browser
... platform-browser-dynamic, router

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.1902.13
@angular-devkit/core         19.2.13
@angular-devkit/schematics   19.2.13
@schematics/angular          20.0.0
ng-packagr                   20.0.0
rxjs                         7.8.2
typescript                   5.8.3
zone.js                      0.15.1
```

### Anything else relevant?

I'm using karma.conf.js `frameworks: ['jasmine', '@angular-devkit/build-angular'],
`

--- a/packages/schematics/angular/migrations/use-application-builder/migration.ts
+++ b/packages/schematics/angular/migrations/use-application-builder/migration.ts
@@ -232,6 +232,8 @@ function updateProjects(tree: Tree, context: SchematicContext) {
 
     // Use @angular/build directly if there is no devkit package usage
     if (!hasAngularDevkitUsage) {
+      const karmaConfigFiles = new Set<string>();
+
       for (const [, target] of allWorkspaceTargets(workspace)) {
         switch (target.builder) {
           case Builders.Application:
@@ -245,9 +247,15 @@ function updateProjects(tree: Tree, context: SchematicContext) {
             break;
           case Builders.Karma:
             target.builder = '@angular/build:karma';
-            // Remove "builderMode" option since the builder will always use "application"
             for (const [, karmaOptions] of allTargetOptions(target)) {
+              // Remove "builderMode" option since the builder will always use "application"
               delete karmaOptions['builderMode'];
+
+              // Collect custom karma configurations for @angular-devkit/build-angular plugin removal
+              const karmaConfig = karmaOptions['karmaConfig'];
+              if (karmaConfig && typeof karmaConfig === 'string') {
+                karmaConfigFiles.add(karmaConfig);
+              }
             }
             break;
           case Builders.NgPackagr:
@@ -291,6 +299,34 @@ function updateProjects(tree: Tree, context: SchematicContext) {
             existing: ExistingBehavior.Replace,
           }),
         );
+      }
+
+      for (const karmaConfigFile of karmaConfigFiles) {
+        if (!tree.exists(karmaConfigFile)) {
+          continue;
+        }
+
+        try {
+          const originalKarmaConfigText = tree.readText(karmaConfigFile);
+          const updatedKarmaConfigText = originalKarmaConfigText
+            .replaceAll(`require('@angular-devkit/build-angular/plugins/karma'),`, '')
+            .replaceAll(`require('@angular-devkit/build-angular/plugins/karma')`, '');
+
+          if (updatedKarmaConfigText.includes('@angular-devkit/build-angular/plugins')) {
+            throw new Error(
+              'Migration does not support found usage of "@angular-devkit/build-angular".',
+            );
+          } else {
+            tree.overwrite(karmaConfigFile, updatedKarmaConfigText);
+          }
+        } catch (error) {
+          const reason = error instanceof Error ? `Reason: ${error.message}` : '';
+          context.logger.warn(
+            `Unable to update custom karma configuration file ("${karmaConfigFile}"). ` +
+              reason +
+              '\nReferences to the "@angular-devkit/build-angular" package within the file may need to be removed manually.',
+          );
+        }
       }
     }
 

--- a/packages/schematics/angular/migrations/use-application-builder/migration_spec.ts
+++ b/packages/schematics/angular/migrations/use-application-builder/migration_spec.ts
@@ -130,6 +130,103 @@ describe(`Migration to use the application builder`, () => {
     expect(builderMode).toBeUndefined();
   });
 
+  it(`should update file for 'karmaConfig' karma option (no require trailing comma)`, async () => {
+    addWorkspaceTarget(tree, 'test', {
+      'builder': Builders.Karma,
+      'options': {
+        'karmaConfig': './karma.conf.js',
+        'polyfills': ['zone.js', 'zone.js/testing'],
+        'tsConfig': 'projects/app-a/tsconfig.spec.json',
+      },
+    });
+    tree.create(
+      './karma.conf.js',
+      `
+        module.exports = function (config) {
+          config.set({
+            basePath: '',
+            frameworks: ['jasmine', '@angular-devkit/build-angular'],
+            plugins: [
+              require('karma-jasmine'),
+              require('karma-chrome-launcher'),
+              require('karma-jasmine-html-reporter'),
+              require('karma-coverage'),
+              require('@angular-devkit/build-angular/plugins/karma')
+            ]
+          });
+        };`,
+    );
+
+    const newTree = await schematicRunner.runSchematic(schematicName, {}, tree);
+    const {
+      projects: { app },
+    } = JSON.parse(newTree.readContent('/angular.json'));
+
+    const { karmaConfig } = app.architect['test'].options;
+    expect(karmaConfig).toBe('./karma.conf.js');
+
+    const karmaConfigText = newTree.readText('./karma.conf.js');
+    expect(karmaConfigText).not.toContain(`require('@angular-devkit/build-angular/plugins/karma')`);
+  });
+
+  it(`should update file for 'karmaConfig' karma option (require trailing comma)`, async () => {
+    addWorkspaceTarget(tree, 'test', {
+      'builder': Builders.Karma,
+      'options': {
+        'karmaConfig': './karma.conf.js',
+        'polyfills': ['zone.js', 'zone.js/testing'],
+        'tsConfig': 'projects/app-a/tsconfig.spec.json',
+      },
+    });
+    tree.create(
+      './karma.conf.js',
+      `
+        module.exports = function (config) {
+          config.set({
+            basePath: '',
+            frameworks: ['jasmine', '@angular-devkit/build-angular'],
+            plugins: [
+              require('karma-jasmine'),
+              require('karma-chrome-launcher'),
+              require('karma-jasmine-html-reporter'),
+              require('karma-coverage'),
+              require('@angular-devkit/build-angular/plugins/karma'),
+            ]
+          });
+        };`,
+    );
+
+    const newTree = await schematicRunner.runSchematic(schematicName, {}, tree);
+    const {
+      projects: { app },
+    } = JSON.parse(newTree.readContent('/angular.json'));
+
+    const { karmaConfig } = app.architect['test'].options;
+    expect(karmaConfig).toBe('./karma.conf.js');
+
+    const karmaConfigText = newTree.readText('./karma.conf.js');
+    expect(karmaConfigText).not.toContain(`require('@angular-devkit/build-angular/plugins/karma')`);
+  });
+
+  it(`should ignore missing file for 'karmaConfig' karma option`, async () => {
+    addWorkspaceTarget(tree, 'test', {
+      'builder': Builders.Karma,
+      'options': {
+        'karmaConfig': './karma.conf.js',
+        'polyfills': ['zone.js', 'zone.js/testing'],
+        'tsConfig': 'projects/app-a/tsconfig.spec.json',
+      },
+    });
+
+    const newTree = await schematicRunner.runSchematic(schematicName, {}, tree);
+    const {
+      projects: { app },
+    } = JSON.parse(newTree.readContent('/angular.json'));
+
+    const { karmaConfig } = app.architect['test'].options;
+    expect(karmaConfig).toBe('./karma.conf.js');
+  });
+
   it('should remove tilde prefix from CSS @import specifiers', async () => {
     // Replace outputPath
     tree.create(

